### PR TITLE
[Refactor] `qmk find`

### PR DESCRIPTION
**Pull request by elpekenin**
## Description

Followup to #21090, move from `if/else` into a `func_name: lambda` mapping and use partial to capture the values.
This reduces some duplication and makes it a little easier to add a new filter.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [X] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: lib/python/qmk/search.py [hunk 1]
```diff
@@ -78,26 +86,17 @@ def search_keymap_targets(keymap='default', filters=[], print_vals=[]):
...
-                            valid_keymaps = filter(lambda e, key=key: key not in e[2], valid_keymaps)
-                        else:
-                            cli.log.warning(f'Unrecognized filter expression: {function_match.group(0)}')
-                            continue
-
-                        cli.log.info(f'Filtering on condition: {{fg_green}}{func_name}{{fg_reset}}({{fg_cyan}}{key}{{fg_reset}})...')
+                    filter_fn = FUNCTIONS.get(func_name)
+                    if filter_fn is None:
+                        cli.log.warning(f'Unrecognized filter expression: {function_match.group(0)}')
+                        continue
```
Comment (sigprof) [inline lib/python/qmk/search.py:92]: The old code also rejected attempts to use functions with a wrong number of arguments; the suggested code does not do that.  One way to fix that could be splitting `FUNCTIONS` into `UNARY_FUNCTIONS` and `BINARY_FUNCTIONS` and choosing one of them here based on `value is not None` (but lambdas in `UNARY_FUNCTIONS` would still need to have a dummy `v` argument, unless you move the call to `functools.partial()` here too).

File: lib/python/qmk/search.py [hunk 2]
```diff
@@ -78,26 +86,17 @@ def search_keymap_targets(keymap='default', filters=[], print_vals=[]):
...
+
+                    # partial because we need to capture current values
+                    # https://github.com/qmk/qmk_firmware/pull/21090
+                    valid_keymaps = filter(functools.partial(filter_fn, k=key, v=value), valid_keymaps)
+
+                    value_text = '' if value is None else f'{{fg_cyan}}{value}{{fg_reset}}'
```
Comment (sigprof) [inline lib/python/qmk/search.py:98]:
```suggestion
                    value_text = '' if value is None else f', {{fg_cyan}}{value}{{fg_reset}}'
```

**Review by sigprof — CHANGES_REQUESTED** [review]

**Review by tzarc — APPROVED** [review]

**Comment by elpekenin** [comment]
Changed the code substantially, existing review does not apply anymore

File: lib/python/qmk/search.py [hunk 3]
```diff
@@ -139,26 +214,14 @@ def _filter_keymap_targets(target_list: List[Tuple[str, str]], filters: List[str
...
-                    if func_name == 'exists':
-                        valid_keymaps = filter(lambda e, key=key: key in e[2], valid_keymaps)
-                    elif func_name == 'absent':
-                        valid_keymaps = filter(lambda e, key=key: key not in e[2], valid_keymaps)
-                    else:
-                        cli.log.warning(f'Unrecognized filter expression: {function_match.group(0)}')
-                        continue
-
-                    cli.log.info(f'Filtering on condition: {{fg_green}}{func_name}{{fg_reset}}({{fg_cyan}}{key}{{fg_reset}})...')
+                filter_class = _get_filter_class(func_name, key, value)
```
Comment (elpekenin) [inline lib/python/qmk/search.py:217]: This code is no longer checking that the binary (key and value) filters receive `value != None`. Shall perhaps add it back

**Review by tzarc — APPROVED** [review]

**Pull request merged**